### PR TITLE
[64137] Split attached table count into attached tables, views and dictionaries

### PR DESCRIPTION
**Pull request by Beetelbrox**
Issue: #64137 

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Change warning on high number of attached tables to differentiate tables, views and dictionaries.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

</details>

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 681de0145888b4dd30d75fd9b1fabe5e2e084b10 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64180/681de0145888b4dd30d75fd9b1fabe5e2e084b10/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by Beetelbrox** [comment]
@Algunenano I've left the warning threshold for both dictionaries & views on 5k, same as tables. You mention in the original issue that this number was arbitrary, do we have any reference about what this number should be for other resources? Or we leave it as-is for the moment?

Thanks!

**Comment by Algunenano** [comment]
> You mention in the original issue that this number was arbitrary, do we have any reference about what this number should be for other resources?

No, I don't think we have anything beyond http://clickhouse.com/docs/knowledgebase/maximum_number_of_tables_and_databases.

I don't have a strong opinion about the numbers, but since views are more lightweight I'd set it to 10k, and since dictionaries live in memory and might have scheduled updates I'd set it to a lower number (1k for example). We can tweak it in the future if we want.

**Comment by Beetelbrox** [comment]
Several fast tests failed, seem completely unrelated tho :thinking:

File: src/Databases/DatabaseLazy.cpp [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #include <Parsers/ASTCreateQuery.h>
 #include <Parsers/ASTFunction.h>
 #include <Storages/IStorage.h>
+#include "Common/CurrentMetrics.h"
```
[the remark this answers is not in the document]
Comment (Beetelbrox) [reply src/Databases/DatabaseLazy.cpp]: Changed quotes & Ordered imports

File: src/Databases/DatabaseLazy.cpp [hunk 2]
```diff
@@ -171,6 +174,22 @@ bool DatabaseLazy::empty() const
     return tables_cache.empty();
 }
 
+static CurrentMetrics::Metric get_attached_count_metric_for_storage(const StoragePtr & storage)
```
[the remark this answers is not in the document]
Comment (Beetelbrox) [reply src/Databases/DatabaseLazy.cpp]: Oops yes, updated.

Didn't find any file that fitted and didn't want to change the IStorage API so moved to a new file `src/Storages/Utils`

**Comment by Beetelbrox** [comment]
> For debugging: https://s3.amazonaws.com/clickhouse-test-reports/64180/24805423544afd3e5c47a736f0da3e47dedac293/fast_test.html -> server.log.zst
> 
> ```
> /tmp $ zstd -d server.log.zst 
> server.log.zst      : 212913589 bytes    
> 
>                                       
> /tmp $ grep "<Fatal>" server.log
> 2024.05.21 20:02:29.833746 [ 246722 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
> 2024.05.21 20:02:29.833842 [ 246722 ] {} <Fatal> BaseDaemon: (version 24.5.1.1, build id: CA71D0AB6C1B16BA5D8120B731ADD1EEC8F12805, git hash: fbe67017e526760167c8c61db02675cdcbd635f6) (from thread 207853) Received signal 11
> 2024.05.21 20:02:29.833899 [ 246722 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
> 2024.05.21 20:02:29.833914 [ 246722 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
> 2024.05.21 20:02:29.833937 [ 246722 ] {} <Fatal> BaseDaemon: Stack trace: 0x00000000073a36e8 0x00000000075d0350
> 2024.05.21 20:02:29.833948 [ 246722 ] {} <Fatal> BaseDaemon: ########################################
> 2024.05.21 20:02:29.834005 [ 246722 ] {} <Fatal> BaseDaemon: (version 24.5.1.1, build id: CA71D0AB6C1B16BA5D8120B731ADD1EEC8F12805, git hash: fbe67017e526760167c8c61db02675cdcbd635f6) (from thread 207853) (no query) Received signal Segmentation fault (11)
> 2024.05.21 20:02:29.834033 [ 246722 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
> 2024.05.21 20:02:29.834065 [ 246722 ] {} <Fatal> BaseDaemon: Stack trace: 0x00000000073a36e8 0x00000000075d0350
> 2024.05.21 20:02:29.848367 [ 246722 ] {} <Fatal> BaseDaemon: 0.0. inlined from /fasttest-workspace/build/./src/Common/StackTrace.cpp:349: StackTrace::tryCapture()
> 2024.05.21 20:02:29.848389 [ 246722 ] {} <Fatal> BaseDaemon: 0. /fasttest-workspace/build/./src/Common/StackTrace.cpp:318: StackTrace::StackTrace(ucontext const&) @ 0x00000000073a36e8
> 2024.05.21 20:02:29.881582 [ 246722 ] {} <Fatal> BaseDaemon: 1. /fasttest-workspace/build/./src/Daemon/BaseDaemon.cpp:0: signalHandler(int, siginfo_t*, void*) @ 0x00000000075d0350
> 2024.05.21 20:02:29.881613 [ 246722 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
> 2024.05.21 20:02:29.928128 [ 246722 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
> 2024.05.21 20:02:34.460903 [ 177158 ] {} <Fatal> Application: Child process was terminated by signal 11.
> 
> /tmp $ grep "\[ 207853 \]" server.log | tail
> 2024.05.21 20:02:27.023494 [ 207853 ] {} <Debug> AsyncLoader: Stop worker in ForegroundLoad
> 2024.05.21 20:02:27.023501 [ 207853 ] {} <Debug> AsyncLoader: Change current priority: 0 -> none
> 2024.05.21 20:02:29.390304 [ 207853 ] {} <Debug> AsyncLoader: Execute load job 'startup Ordinary database test_x9zix6bq' in ForegroundLoad
> 2024.05.21 20:02:29.390331 [ 207853 ] {} <Debug> AsyncLoader: Finish load job 'startup Ordinary database test_x9zix6bq' with status OK
> 2024.05.21 20:02:29.390343 [ 207853 ] {} <Debug> AsyncLoader: Spawn loader worker #2 in ForegroundLoad
> 2024.05.21 20:02:29.390362 [ 207853 ] {} <Debug> AsyncLoader: Execute load job 'startup Atomic database test_x9zix6bq' in ForegroundLoad
> 2024.05.21 20:02:29.390396 [ 207853 ] {} <Debug> AsyncLoader: Finish load job 'startup Atomic database test_x9zix6bq' with status OK
> 2024.05.21 20:02:29.390410 [ 207853 ] {} <Debug> AsyncLoader: Stop worker in ForegroundLoad
> 2024.05.21 20:02:29.390421 [ 207853 ] {} <Debug> AsyncLoader: Change current priority: 0 -> none
> 2024.05.21 20:02:29.833204 [ 207853 ] {} <Debug> DatabaseLazy (lazy_test_p8receso): Attach table `таблица_со_странным_названием`.
> ```
> 
> That should help pinpoint the test and the issue

Yes, thanks! Apolog

> For debugging: https://s3.amazonaws.com/clickhouse-test-reports/64180/24805423544afd3e5c47a736f0da3e47dedac293/fast_test.html -> server.log.zst
> 
> ```
> /tmp $ zstd -d server.log.zst 
> server.log.zst      : 212913589 bytes    
> 
>                                       
> /tmp $ grep "<Fatal>" server.log
> 2024.05.21 20:02:29.833746 [ 246722 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
> 2024.05.21 20:02:29.833842 [ 246722 ] {} <Fatal> BaseDaemon: (version 24.5.1.1, build id: CA71D0AB6C1B16BA5D8120B731ADD1EEC8F12805, git hash: fbe67017e526760167c8c61db02675cdcbd635f6) (from thread 207853) Received signal 11
> 2024.05.21 20:02:29.833899 [ 246722 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
> 2024.05.21 20:02:29.833914 [ 246722 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
> 2024.05.21 20:02:29.833937 [ 246722 ] {} <Fatal> BaseDaemon: Stack trace: 0x00000000073a36e8 0x00000000075d0350
> 2024.05.21 20:02:29.833948 [ 246722 ] {} <Fatal> BaseDaemon: ########################################
> 2024.05.21 20:02:29.834005 [ 246722 ] {} <Fatal> BaseDaemon: (version 24.5.1.1, build id: CA71D0AB6C1B16BA5D8120B731ADD1EEC8F12805, git hash: fbe67017e526760167c8c61db02675cdcbd635f6) (from thread 207853) (no query) Received signal Segmentation fault (11)
> 2024.05.21 20:02:29.834033 [ 246722 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
> 2024.05.21 20:02:29.834065 [ 246722 ] {} <Fatal> BaseDaemon: Stack trace: 0x00000000073a36e8 0x00000000075d0350
> 2024.05.21 20:02:29.848367 [ 246722 ] {} <Fatal> BaseDaemon: 0.0. inlined from /fasttest-workspace/build/./src/Common/StackTrace.cpp:349: StackTrace::tryCapture()
> 2024.05.21 20:02:29.848389 [ 246722 ] {} <Fatal> BaseDaemon: 0. /fasttest-workspace/build/./src/Common/StackTrace.cpp:318: StackTrace::StackTrace(ucontext const&) @ 0x00000000073a36e8
> 2024.05.21 20:02:29.881582 [ 246722 ] {} <Fatal> BaseDaemon: 1. /fasttest-workspace/build/./src/Daemon/BaseDaemon.cpp:0: signalHandler(int, siginfo_t*, void*) @ 0x00000000075d0350
> 2024.05.21 20:02:29.881613 [ 246722 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
> 2024.05.21 20:02:29.928128 [ 246722 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
> 2024.05.21 20:02:34.460903 [ 177158 ] {} <Fatal> Application: Child process was terminated by signal 11.
> 
> /tmp $ grep "\[ 207853 \]" server.log | tail
> 2024.05.21 20:02:27.023494 [ 207853 ] {} <Debug> AsyncLoader: Stop worker in ForegroundLoad
> 2024.05.21 20:02:27.023501 [ 207853 ] {} <Debug> AsyncLoader: Change current priority: 0 -> none
> 2024.05.21 20:02:29.390304 [ 207853 ] {} <Debug> AsyncLoader: Execute load job 'startup Ordinary database test_x9zix6bq' in ForegroundLoad
> 2024.05.21 20:02:29.390331 [ 207853 ] {} <Debug> AsyncLoader: Finish load job 'startup Ordinary database test_x9zix6bq' with status OK
> 2024.05.21 20:02:29.390343 [ 207853 ] {} <Debug> AsyncLoader: Spawn loader worker #2 in ForegroundLoad
> 2024.05.21 20:02:29.390362 [ 207853 ] {} <Debug> AsyncLoader: Execute load job 'startup Atomic database test_x9zix6bq' in ForegroundLoad
> 2024.05.21 20:02:29.390396 [ 207853 ] {} <Debug> AsyncLoader: Finish load job 'startup Atomic database test_x9zix6bq' with status OK
> 2024.05.21 20:02:29.390410 [ 207853 ] {} <Debug> AsyncLoader: Stop worker in ForegroundLoad
> 2024.05.21 20:02:29.390421 [ 207853 ] {} <Debug> AsyncLoader: Change current priority: 0 -> none
> 2024.05.21 20:02:29.833204 [ 207853 ] {} <Debug> DatabaseLazy (lazy_test_p8receso): Attach table `таблица_со_странным_названием`.
> ```
> 
> That should help pinpoint the test and the issue

Thanks! I'd reached the segfault but wasn't being able to pinpoint the test. Test is `01015_database_bad_tables`, segfault reproduced locally.

Turns out that with the `Lazy` engine (`DatabaseLazy::attach/detach`) the `StoragePtr` can be arbitrarily null so we can't dereference it. Nevertheless the [docs](https://clickhouse.com/docs/en/engines/database-engines/lazy) say that the lazy engine can only be used with `Log` tables, I've tested it and seems to be the case:

```
fjjm-laptop :) CREATE DATABASE teste ENGINE = Lazy(1);

CREATE DATABASE teste
ENGINE = Lazy(1)

Query id: 87827036-c63d-4d4d-a7ca-f83d65e54a4c

Ok.

0 rows in set. Elapsed: 0.002 sec. 

fjjm-laptop :) CREATE TABLE teste.log_test (id Int32, str String) Engine=Log

CREATE TABLE teste.log_test
(
    `id` Int32,
    `str` String
)
ENGINE = Log

Query id: 7babc873-5c0b-4611-a616-3c42294c92b9

Ok.

0 rows in set. Elapsed: 0.002 sec. 

fjjm-laptop :) CREATE VIEW teste.log_test_view AS select * FROM teste.log_test

CREATE VIEW teste.log_test_view
AS SELECT *
FROM teste.log_test

Query id: 9241b8b9-6824-4f64-87dc-7d61f881a62b


Elapsed: 0.003 sec. 

Received exception from server (version 24.5.1):
Code: 1. DB::Exception: Received from localhost:9000. DB::Exception: Lazy engine can be used only with *Log tables.. (UNSUPPORTED_METHOD)

fjjm-laptop :) CREATE DICTIONARY teste.log_dict (id Int32, str String) PRIMARY KEY id SOURCE(CLICKHOUSE(db 'teste' TABLE 'log_test')) LAYOUT(FLAT()) LIFETIME(MIN 0 MAX 1000)

CREATE DICTIONARY teste.log_dict
(
    `id` Int32,
    `str` String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(DB 'teste' TABLE 'log_test'))
LIFETIME(MIN 0 MAX 1000)
LAYOUT(FLAT())

Query id: 4093360b-ca6b-4507-95d6-85b053b53800


Elapsed: 0.003 sec. 

Received exception from server (version 24.5.1):
Code: 1. DB::Exception: Received from localhost:9000. DB::Exception: Lazy engine can be used only with *Log tables.. (UNSUPPORTED_METHOD)
```

So I'm reverting the counting to only take into account tables so we don't need the `StoragePtr` in lazy DBs

**Pull request merged**